spring-kafka/src/main/java/org/springframework/kafka/ReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,304 @@
+
+	private TaskScheduler scheduler = new ThreadPoolTaskScheduler();
+
+	private boolean running;
+
+	private int phase;
+
+	private boolean autoStartup;
+
+	private long replyTimeout = DEFAULT_REPLY_TIMEOUT;
+
+	private volatile boolean schedulerSet;
+
+	public ReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer) {
+		this(producerFactory, replyContainer, false);
+	}
+
+	public ReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer, boolean autoFlush) {
+		super(producerFactory, autoFlush);
+		Assert.notNull(replyContainer, "'replyContainer' cannot be null");
+		this.replyContainer = replyContainer;
+		this.replyContainer.setupMessageListener(this);
+	}
+
+	protected void setTaskScheduler(TaskScheduler scheduler) {
+		this.scheduler = scheduler;
+		this.schedulerSet = true;
+	}
+
+	protected void setReplyTimeout(long replyTimeout) {
+		Assert.isTrue(replyTimeout >= 0, "'replyTimeout' must be >= 0");
+		this.replyTimeout = replyTimeout;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		if (!this.schedulerSet) {
+			((ThreadPoolTaskScheduler) this.scheduler).initialize();
+		}
+	}
+
+	@Override
+	public synchronized void start() {
+		try {
+			afterPropertiesSet();
+		}
+		catch (Exception e) {
+			throw new KafkaException("Failed to initialize", e);
+		}
+		this.replyContainer.start();
+		this.running = true;
+	}
+
+	@Override
+	public synchronized void stop() {
+		this.running = false;
+		this.replyContainer.stop();
+	}
+
+	@Override
+	public synchronized boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public RequestReplyFuture<K, V, R> sendAndReceive(ProducerRecord<K, V> record) {
+		CorrelationKey correlationId = createCorrelationId();
+		record.headers().add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.correlationId));
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("Sending: " + record + " with correlationId: " + correlationId);
+		}
+		TemplateRequestReplyFuture<K, V, R> future = new TemplateRequestReplyFuture<>();
+		this.futures.put(correlationId, future);
+		future.setSendFuture(send(record));
+		this.scheduler.schedule(() -> {
+			RequestReplyFuture<K, V, R> removed = this.futures.remove(correlationId);
+			if (removed != null) {
+				if (this.logger.isWarnEnabled()) {
+					this.logger.warn("Reply timed out for: " + record + " with correlationId: " + correlationId);
```

Comment:
When we call `stop()` we perform `this.replyContainer.stop()` therefore no any replies are going to come back to us.
And from here all the waiting `this.futures` will be discarded here with this misleading warning.

What I think we need to cancel the `Future` from this `schedule()` and clean up `this.futures` altogether in the `stop()` as well after stopping `this.replyContainer`.

WDYT?

Reply:
Canceling the futures would mean we'd have to track them. I think we can just do

    this.futures.replaceAll((k, v) -> null);

in `stop()` and then the timeouts will be no-ops.

Reply:
Actually CHM doesn't support `null`, but `clear()` will do it.